File: backbone-python/examples/clean_api_python/application/commands/update_product.py
"""UpdateProduct command + handler — Application / Commands"""
from dataclasses import dataclass
from typing import Dict, Any, Optional
import sys, os

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../../..')))

from backbone.infrastructure.logging import LoggerFactory
from examples.clean_api_python.domain.entities.product import ValidationError
from examples.clean_api_python.domain.repositories.product_repository import ProductRepository


@dataclass
class UpdateProductCommand:
    id: str
    name: Optional[str] = None
    description: Optional[str] = None
    price: Optional[float] = None
    category: Optional[str] = None
    stock: Optional[int] = None


@dataclass
class UpdateProductResult:
    product_id: str


class UpdateProductCommandHandler:

    def __init__(self, repository: ProductRepository):
        self.repository = repository
        self.logger = LoggerFactory.create_layer_logger(
            "product-api", "application", "UpdateProductCommandHandler"
        )

    def handle(self, command: UpdateProductCommand, context: Dict[str, Any]) -> UpdateProductResult:
        self.logger.info("Handling UpdateProductCommand", extra_data={"product_id": command.id})

        product = self.repository.find_by_id(command.id)
        if product is None:
            raise ValueError(f"Product {command.id} not found")

        if command.name is not None:
            product.name = command.name
        if command.description is not None:
            product.description = command.description
        if command.category is not None:
            product.category = command.category
        if command.price is not None:
            try:
                product.update_price(command.price)
            except ValidationError as e:
                raise ValueError(e.message)
        if command.stock is not None:
            try:
                product.update_stock(command.stock)
            except ValidationError as e:
                raise ValueError(e.message)

        try:
            product.validate()
        except ValidationError as e:
            raise ValueError(e.message)

        self.repository.update(product)
        self.logger.info("Product updated", extra_data={"product_id": product.id})

        return UpdateProductResult(product_id=product.id)

File: backbone-go/examples/clean-api-go/application/commands/update_product.go
// Package commands contains write-side application handlers (CQRS).
package commands

import (
	"context"
	"fmt"
	"time"

	"github.com/freakjazz/backbone-go/examples/clean-api-go/domain/repositories"
	"github.com/freakjazz/backbone-go/infrastructure/logging"
)

// UpdateProductCommand is the write input for updating a product.
type UpdateProductCommand struct {
	ID          string
	Name        string
	Description string
	Price       float64
	Category    string
	Stock       int
}

// UpdateProductResult is the write output.
type UpdateProductResult struct {
	ProductID string
}

// UpdateProductCommandHandler handles the UpdateProduct command.
type UpdateProductCommandHandler struct {
	repository repositories.ProductRepository
	logger     *logging.EnhancedLogger
}

// NewUpdateProductCommandHandler creates a new handler instance.
func NewUpdateProductCommandHandler(repo repositories.ProductRepository, logger *logging.EnhancedLogger) *UpdateProductCommandHandler {
	return &UpdateProductCommandHandler{repository: repo, logger: logger}
}

// Handle executes the command.
func (h *UpdateProductCommandHandler) Handle(ctx context.Context, cmd UpdateProductCommand) (*UpdateProductResult, error) {
	log := h.logger.
		WithLayer("application").
		WithHandler("UpdateProductCommandHandler").
		WithMethod("Handle").
		WithContext(map[string]interface{}{"request_id": ctx.Value("request_id"), "product_id": cmd.ID})

	if cmd.ID == "" {
		return nil, fmt.Errorf("product ID is required")
	}

	product, err := h.repository.FindByID(ctx, cmd.ID)
	if err != nil {
		log.Warning("Product not found", map[string]interface{}{"error": err.Error()})
		return nil, err
	}

	if cmd.Name != "" {
		product.Name = cmd.Name
	}
	if cmd.Description != "" {
		product.Description = cmd.Description
	}
	if cmd.Category != "" {
		product.Category = cmd.Category
	}
	if cmd.Price != 0 {
		if err := product.UpdatePrice(cmd.Price); err != nil {
			return nil, err
		}
	}
	if cmd.Stock >= 0 {
		if err := product.UpdateStock(cmd.Stock); err != nil {
			return nil, err
		}
	}

	if err := product.Validate(); err != nil {
		log.ErrorWithCode("Domain validation failed", 10003001, map[string]interface{}{"error": err.Error()})
		return nil, err
	}

	start := time.Now()
	if err := h.repository.Update(ctx, product); err != nil {
		log.ErrorWithCode("Failed to persist update", 10003002, map[string]interface{}{
			"error": err.Error(), "duration_ms": time.Since(start).Milliseconds(),
		})
		return nil, err
	}

	log.Info("Product updated", map[string]interface{}{
		"product_id": product.ID, "duration_ms": time.Since(start).Milliseconds(),
	})
	return &UpdateProductResult{ProductID: product.ID}, nil
}
